File: Programmers/Level_3/가장긴팰린드롬.py
'''
https://programmers.co.kr/learn/courses/30/lessons/12904

의외로 2중 for문으로 풀리는 문제
시간 복잡도 고려를 잘못해서 2중 for문으로 풀면 안 되는줄 알았다
2500!이 아니라 2500+2499+2498+...1 까지로 NM에 가까운 문제였다
고로 문자열을 i~j까지 슬라이싱 해서 해당 부분 문자열이 팰린드롬인지 매번 체크해도 된다
'''

def solution(s):
    answer = 1
    
    def palindrome(substring):
        if substring == substring[::-1]:
            return True
        
    for i in range(len(s)):
        for j in range(i+1, len(s)):
            if palindrome(s[i:j+1]):
                answer = max(answer, len(s[i:j+1]))
    
    return answer

print(solution("abcdcba"))

File: Programmers/Level_3/Swift_Code/Swift_Code/가장긴팰린드롬.swift
//
//  가장긴팰린드롬.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/03/25.
//

/*
 파이썬과 동일한 로직으로 수행 시 현재 코드보다 3배 가량 오래 걸림
 하지만 현재 코드도 효율성 2개 중 1개에서 걸림
 Swift로 통과하려면 로직 전체를 수정해야할 필요가 있지만
 파이썬 로직이 틀린 것은 아니므로 여기까지
 */

import Foundation

extension String {
    var isPalindrome: Bool {
        self == String(self.reversed())
    }
}

func solution(_ s:String) -> Int {
    var answer = 1
    
    for i in 0..<s.count {
        for j in stride(from: s.count - i, to: answer, by: -1) {
            let startIndex = s.index(s.startIndex, offsetBy: i)
            let endIndex = s.index(s.startIndex, offsetBy: i + j - 1)
            let substring = String(s[startIndex...endIndex])
            if substring.isPalindrome {
                answer = max(answer, substring.count)
            }
        }
    }
    
    return answer
}

// print(solution("abcdcba"))
